### PR TITLE
Add history

### PR DESCRIPTION
**Pull request by vbragin**

### CODE REVIEW

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 1]
```diff
@@ -98,17 +107,80 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+        copyHistory(artifactsTempDir, chain, chainExecution);
         addExecutorInfo(artifactsTempDir, chain, chainExecution);
     }
 
+    /**
+     * Write the history file to results directory.
+     */
+    private void copyHistory(File artifactsTempDir, Chain chain, ChainExecution chainExecution) {
+        try {
+            File historyDir = new File(artifactsTempDir.toPath().resolve("history").toString());
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: `Path historyDir = artifactsTempDir.toPath().resplve("history");`

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 2]
```diff
@@ -98,17 +107,80 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+            if(!historyDir.mkdir()) {
+                return;
+            }
+            boolean historyFilesNotFound = true;
+            int buildNumber = chainExecution.getPlanResultKey().getBuildNumber();
+            while (historyFilesNotFound) {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: find build first, then copy artifacts

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 3]
```diff
@@ -98,17 +107,80 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+                ResultsSummary lastBuild = Optional.ofNullable(resultsSummaryManager.findLastBuildResultBefore(
+                        chain.getPlanKey().getKey(), buildNumber)).orElseThrow(
+                        () -> new RuntimeException("Could not find previous buildId"));
+                if (historyArtifactExists(chain.getPlanKey().getKey(), lastBuild.getBuildNumber())) {
+                    copyArtifactToHistoryFolder(historyDir.toPath(), "history.json", chain.getPlanKey().getKey(), lastBuild.getBuildNumber());
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: store file names to const, then iterate over all the files

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 4]
```diff
@@ -98,17 +107,80 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+                    buildNumber = lastBuild.getBuildNumber();
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.info("Cannot copy history files.", e);
+        }
+    }
+
+    private boolean historyArtifactExists(String planKey, int buildId) {
+        String artifactUrl= getHistoryArtifactUrl("history.json", planKey, buildId);
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: indents

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 5]
```diff
@@ -98,17 +107,80 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+        } catch (Exception e) {
+            LOGGER.info("Cannot connect to artifact {}.", artifactUrl, e);
+            return false;
+        }
+    }
+
+    private void copyArtifactToHistoryFolder(Path historyFolder, String fileName, String planKey, int buildId) throws Exception {
+        URL artifactUrl = new URL(getHistoryArtifactUrl(fileName, planKey, buildId));
+        URLConnection uc = artifactUrl.openConnection();
+        InputStream inputStream = uc.getInputStream();
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: try-with-resources

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 6]
```diff
@@ -98,17 +107,84 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+        } catch (Exception e) {
+            LOGGER.info("Cannot copy history files.", e);
+        }
+    }
+
+    private Integer getLastBuildNumberWithHistory(String planKey, int buildNumber) {
+        while (true) {
+            ResultsSummary lastBuild = Optional.ofNullable(resultsSummaryManager.findLastBuildResultBefore(
+                    planKey, buildNumber)).orElseThrow(
+                    () -> new RuntimeException("Could not found a previous buildId with the history artifact."));
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: You can return `Optional.empty` instead

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 7]
```diff
@@ -98,17 +107,84 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
+            LOGGER.info("Cannot connect to artifact {}.", artifactUrl, e);
+            return false;
+        }
+    }
+
+    private void copyArtifactToHistoryFolder(Path historyFolder, String fileName, String planKey, int buildId) throws Exception {
+        URL artifactUrl = new URL(getHistoryArtifactUrl(fileName, planKey, buildId));
+        URLConnection uc = artifactUrl.openConnection();
+        try(InputStream inputStream = uc.getInputStream()) {
+            if (!historyFolder.toFile().exists()) {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: `Files.createDirectories()`

File: src/main/java/io/qameta/allure/bamboo/AllureReportServlet.java [hunk 8]
```diff
@@ -66,21 +66,55 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
+    protected void doHead(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException {
+        final Matcher matcher = URL_PATTERN.matcher(request.getRequestURI());
+        if (matcher.matches()) {
+            response.setHeader("X-Frame-Options", "ALLOWALL");
+            final String planKeyString = matcher.group(1);
+            final String buildNumber = matcher.group(2);
+            final String filePath = matcher.group(3);
+            final PlanResultKey planResultKey = getPlanResultKey(planKeyString, parseInt(buildNumber));
+            ofNullable(resultsSummaryManager.getResultsSummary(planResultKey)).ifPresent(results -> {
```
Comment (baev) [inline src/main/java/io/qameta/allure/bamboo/AllureReportServlet.java]: can we reduce complexity of this method?

Comment (vbragin) [reply src/main/java/io/qameta/allure/bamboo/AllureReportServlet.java]: In later releases.

File: src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java [hunk 9]
```diff
@@ -98,17 +111,89 @@ public void execute(@NotNull Chain chain, @NotNull ChainResultsSummary chainResu
...
-        new AddExecutorInfo(rootUrl, buildId, buildName, buildUrl, reportUrl).invoke(artifactsTempDir);
+        new AddExecutorInfo(rootUrl, Integer.toString(buildId), buildName, buildUrl, reportUrl).invoke(artifactsTempDir);
+    }
+
+    /**
+     * Returns the base url of bamboo server.
+     */
+    @NotNull
+    private String getBambooBaseUrl() {
+        return getAdministrationConfiguration().getBaseUrl();
```
Comment (smecsia) [inline src/main/java/io/qameta/allure/bamboo/AllureBuildCompleteAction.java]: Are you sure this always returns non-null and correct value?

**Comment by smecsia** [comment]
Generally I'm fine with the changes, so this is approved.

**Pull request merged**